src/dfx-core/src/config/cache.rs

```diff
@@ -17,7 +17,7 @@ pub trait Cache {
     fn get_binary_command(&self, binary_name: &str) -> Result<std::process::Command, CacheError>;
 }
 
-pub fn get_cache_root() -> Result<PathBuf, CacheError> {
+pub fn get_cache_root(ensure_dir_exists: bool) -> Result<PathBuf, CacheError> {
```

Comment:
This approach doesn't work for at least these reasons:
- booleans aren't the most readable type to pass around as parameters
- this particular boolean is a lie at different levels:
  - get_cache_root(ensure_dir_exists = true): ensures `<cache root>` exists, returns `<cache root> ...  ok
  - get_bin_cache_root(ensure_dir_exists = true): ensures `<cache root>/versions` exists, returns `<cache root>/versions` .. ok
  - get_bin_cache(version: &str, ensure_dir_exists) = true: ensures `<cache root>/versions` exists, returns `<cache root>/versions/<dfx version>` .. oops

Luckily, `get_cache_root` only has two callers, and one of them adds two levels of subdirectory.

Let's try this:
- remove all these `ensure_dir_exists : bool` parameters
- change `get_cache_root` to construct and return the path without creating any directories
- add a new method 'get_cache_path_for_version(v: &str)` that constructs and returns `<cache root>/versions/<version>` without creating any directories
- use `get_cache_path_for_version(v: &str)` to determine if there are extensions, without having to create the cache directory.

Reply:
very nice